src/modules/0.0.7/aws_ec2_metadata/index.ts

```diff
@@ -12,20 +12,25 @@ export const AwsEc2MetadataModule: Module2 = new Module2({
     instanceMetadataMapper: async (instance: AWSInstance, ctx: Context) => {
       const client = await ctx.getAwsClient() as AWS;
       const out = new InstanceMetadata();
-      out.instanceId = instance.InstanceId ?? '';
+      if (!instance.InstanceId) return undefined;
+      out.instanceId = instance.InstanceId;
       // fill join column which is the id from the `instance` table
       const ins = await AwsEc2Module.mappers.instance.db.read(
         ctx,
         out.instanceId,
       );
       out.id = ins.id;
       out.architecture = instance.Architecture as Architecture;
-      out.privateIpAddress = instance.PrivateIpAddress ?? '';
+      if (!instance.PrivateIpAddress) return undefined;
+      out.privateIpAddress = instance.PrivateIpAddress;
       out.launchTime = instance.LaunchTime as Date;
       out.spot = instance.InstanceLifecycle === 'spot';
-      out.cpuCores = instance.CpuOptions?.CoreCount as number;
-      const instanceType = await client.getInstanceType(instance.InstanceType ?? '');
-      out.memSizeMB = instanceType?.MemoryInfo?.SizeInMiB as number;
+      if (!instance.CpuOptions?.CoreCount) return undefined;
+      out.cpuCores = instance.CpuOptions?.CoreCount;
+      if (!instance.InstanceType) return undefined;
+      const instanceType = await client.getInstanceType(instance.InstanceType);
+      if (!instanceType?.MemoryInfo?.SizeInMiB) return undefined;
+      out.memSizeMB = instanceType?.MemoryInfo?.SizeInMiB;
```

Comment:
nit: probably would make sense to have a "precondition" check at the beginning of the mapper checking if any of the `required` properties is wrong just `return undefined;` once.

Reply:
Also, I agree replacing `throw`s and return `undefined`s instead, but I'm not entirely sure if in this case it should be the approach, I guess if we are 100% sure these properties always have to have a valid value it makes sense but if any of them could not be there we could default to something